x/merkledb/view.go

```diff
@@ -290,12 +290,66 @@ func (v *view) hashChangedNodes(ctx context.Context) {
 // Calculates the ID of all descendants of [n] which need to be recalculated,
 // and then calculates the ID of [n] itself.
 func (v *view) hashChangedNode(n *node) ids.ID {
-	// We use [wg] to wait until all descendants of [n] have been updated.
-	var wg sync.WaitGroup
+	// If there are no children, we can avoid allocating [keyBuffer].
+	if len(n.children) == 0 {
+		return n.calculateID(v.db.metrics)
+	}
+
+	// Calculate the size of the largest child key of this node. This allows
+	// only allocating a single slice for all of the keys.
+	var maxChildBitLength int
+	for _, childEntry := range n.children {
+		maxChildBitLength = max(maxChildBitLength, childEntry.compressedKey.length)
+	}
+
+	var (
+		maxBytesNeeded = bytesNeeded(n.key.length + v.tokenSize + maxChildBitLength)
+		// keyBuffer is allocated onto the heap because it is dynamically sized.
+		keyBuffer = make([]byte, maxBytesNeeded)
+		// childBuffer is allocated on the stack.
+		childBuffer = make([]byte, 1)
+		dualIndex   = dualBitIndex(v.tokenSize)
+		bytesForKey = bytesNeeded(n.key.length)
+		// We track the last byte of [n.key] so that we can reset the value for
+		// each key. This is needed because the child buffer may get ORed at
+		// this byte.
+		lastKeyByte byte
+
+		// We use [wg] to wait until all descendants of [n] have been updated.
+		wg sync.WaitGroup
+	)
+
+	if bytesForKey > 0 {
+		// We can just copy this node's key once. It doesn't change as we
+		// iterate over the children.
```

Comment:
nit: this wording feels a bit too informal with "just", maybe this?

```suggestion
		// We only need to copy this node's key once because it does not change as we iterate over the children.
```